File: go/processor/extractor/observability.go
// @docspec:module {
//   id: "docspec-go-extractor-observability",
//   name: "Observability Extractor",
//   description: "Detects Prometheus metrics, OpenTelemetry instrumentation, DataDog statsd, and health check endpoints to populate the observability section.",
//   since: "3.0.0"
// }
package extractor

import (
	"go/ast"
	"strings"
)

// ObservabilityExtractor detects observability instrumentation in Go source code:
// @docspec:boundary "classpath-safe extraction"
//   - Prometheus metrics (prometheus.NewCounter, prometheus.NewHistogram, etc.)
//   - OpenTelemetry tracing (otel tracer usage)
//   - Health check endpoints (common patterns in Gin, Echo, net/http)
//   - Comment-based docspec tags (//docspec:metric, //docspec:healthcheck)
//
// This is the Go equivalent of the Java ObservabilityExtractor which detects
// Micrometer @Timed/@Counted and Spring Actuator HealthIndicator.
type ObservabilityExtractor struct{}

// Known Go observability import paths.
var observabilityImports = []string{
	"github.com/prometheus/client_golang/prometheus",
	"github.com/prometheus/client_golang/prometheus/promauto",
	"go.opentelemetry.io/otel",
	"go.opentelemetry.io/otel/trace",
	"go.opentelemetry.io/otel/metric",
	"github.com/DataDog/datadog-go/statsd",
	"github.com/armon/go-metrics",
}

// @docspec:deterministic
func (e *ObservabilityExtractor) Name() string {
	return "observability"
}

// IsAvailable reports true if any file imports a known observability package.
//
// @docspec:intentional "Checks observability library imports and docspec:metric/docspec:healthcheck tags"
func (e *ObservabilityExtractor) IsAvailable(ctx *ProcessorContext) bool {
	for _, imp := range observabilityImports {
		if hasImport(ctx, imp) {
			return true
		}
	}
	// Also check for docspec observability tags
	for _, file := range ctx.Files {
		if hasObservabilityDocTags(file) {
			return true
		}
	}
	return false
}

// Extract scans files for observability patterns and populates ctx.Observability.
//
// @docspec:method { since: "3.0.0" }
// @docspec:intentional "Discovers Prometheus NewCounter/NewHistogram declarations, OTel meter calls, health check handlers, and docspec:metric/docspec:healthcheck tags"
func (e *ObservabilityExtractor) Extract(ctx *ProcessorContext) {
	var metrics []ObservabilityMetric
	var healthChecks []ObservabilityHealthCheck
	seen := make(map[string]bool)

	for _, file := range ctx.Files {
		imports := fileImports(file)
		pkg := file.Name.Name

		hasProm := imports["github.com/prometheus/client_golang/prometheus"] ||
			imports["github.com/prometheus/client_golang/prometheus/promauto"]
		hasOtel := imports["go.opentelemetry.io/otel/metric"]

		for _, decl := range file.Decls {
			switch d := decl.(type) {
			case *ast.GenDecl:
				// Detect global var declarations like:
				// var requestCount = prometheus.NewCounter(...)
				// var requestDuration = promauto.NewHistogram(...)
				for _, spec := range d.Specs {
					vs, ok := spec.(*ast.ValueSpec)
					if !ok || len(vs.Values) == 0 {
						continue
					}
					for i, val := range vs.Values {
						call, ok := val.(*ast.CallExpr)
						if !ok {
							continue
						}
						name, metricType := extractPrometheusMetric(call, hasProm)
						if name == "" {
							continue
						}
						if seen[name] {
							continue
						}
						seen[name] = true

						varName := ""
						if i < len(vs.Names) {
							varName = vs.Names[i].Name
						}

						m := ObservabilityMetric{
							Name:      name,
							Type:      metricType,
							EmittedBy: []string{pkg + "." + varName},
						}
						labels := extractPrometheusLabels(call)
						if len(labels) > 0 {
							m.Labels = labels
						}
						metrics = append(metrics, m)
					}
				}

			case *ast.FuncDecl:
				if !d.Name.IsExported() {
					continue
				}

				owner := pkg + "." + d.Name.Name
				if d.Recv != nil && len(d.Recv.List) > 0 {
					owner = pkg + "." + recvTypeName(d.Recv.List[0].Type) + "." + d.Name.Name
				}

				// Check docspec comment tags
				if d.Doc != nil {
					for _, c := range d.Doc.List {
						text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))

						if strings.HasPrefix(text, "docspec:metric") {
							parts := strings.Fields(text)
							if len(parts) >= 3 {
								name := parts[1]
								mtype := parts[2]
								if !seen[name] {
									seen[name] = true
									metrics = append(metrics, ObservabilityMetric{
										Name:      name,
										Type:      mtype,
										EmittedBy: []string{owner},
									})
								}
							}
						}

						if strings.HasPrefix(text, "docspec:healthcheck") {
							parts := strings.SplitN(text, " ", 2)
							path := "/health"
							if len(parts) == 2 {
								path = strings.TrimSpace(parts[1])
							}
							healthChecks = append(healthChecks, ObservabilityHealthCheck{
								Path:   path,
								Checks: []string{d.Name.Name},
							})
						}
					}
				}

				// Detect OTel meter calls inside function bodies
				if hasOtel && d.Body != nil {
					ast.Inspect(d.Body, func(n ast.Node) bool {
						call, ok := n.(*ast.CallExpr)
						if !ok {
							return true
						}
						sel, ok := call.Fun.(*ast.SelectorExpr)
						if !ok {
							return true
						}
						metricType := ""
						switch sel.Sel.Name {
						case "Int64Counter", "Float64Counter":
							metricType = "counter"
						case "Int64Histogram", "Float64Histogram":
							metricType = "histogram"
						case "Int64UpDownCounter", "Float64UpDownCounter":
							metricType = "gauge"
						}
						if metricType != "" && len(call.Args) > 0 {
							if lit, ok := call.Args[0].(*ast.BasicLit); ok {
								name := strings.Trim(lit.Value, `"`)
								if !seen[name] {
									seen[name] = true
									metrics = append(metrics, ObservabilityMetric{
										Name:      name,
										Type:      metricType,
										EmittedBy: []string{owner},
									})
								}
							}
						}
						return true
					})
				}

				// Detect health check handler patterns
				if d.Body != nil {
					if isHealthCheckFunc(d) {
						healthChecks = append(healthChecks, ObservabilityHealthCheck{
							Path:   "/health",
							Checks: []string{d.Name.Name},
						})
					}
				}
			}
		}
	}

	if len(metrics) == 0 && len(healthChecks) == 0 {
		return
	}

	if ctx.Observability == nil {
		ctx.Observability = &ObservabilityModel{}
	}
	ctx.Observability.Metrics = append(ctx.Observability.Metrics, metrics...)
	ctx.Observability.HealthChecks = append(ctx.Observability.HealthChecks, healthChecks...)
}

// extractPrometheusMetric extracts metric name and type from a prometheus.NewXxx call.
//
// @docspec:deterministic
func extractPrometheusMetric(call *ast.CallExpr, hasProm bool) (string, string) {
	if !hasProm {
		return "", ""
	}

	sel, ok := call.Fun.(*ast.SelectorExpr)
	if !ok {
		return "", ""
	}

	metricType := ""
	switch sel.Sel.Name {
	case "NewCounter", "NewCounterVec":
		metricType = "counter"
	case "NewGauge", "NewGaugeVec":
		metricType = "gauge"
	case "NewHistogram", "NewHistogramVec":
		metricType = "histogram"
	case "NewSummary", "NewSummaryVec":
		metricType = "summary"
	default:
		return "", ""
	}

	// Extract name from the Opts struct literal argument
	if len(call.Args) > 0 {
		comp, ok := call.Args[0].(*ast.CompositeLit)
		if ok {
			name := extractStructField(comp, "Name")
			if name != "" {
				namespace := extractStructField(comp, "Namespace")
				subsystem := extractStructField(comp, "Subsystem")
				fullName := joinMetricName(namespace, subsystem, name)
				return fullName, metricType
			}
		}
		// Unary address might wrap the struct
		unary, ok := call.Args[0].(*ast.UnaryExpr)
		if ok {
			comp, ok := unary.X.(*ast.CompositeLit)
			if ok {
				name := extractStructField(comp, "Name")
				if name != "" {
					namespace := extractStructField(comp, "Namespace")
					subsystem := extractStructField(comp, "Subsystem")
					fullName := joinMetricName(namespace, subsystem, name)
					return fullName, metricType
				}
			}
		}
	}

	return "", ""
}

// extractPrometheusLabels extracts label names from a Vec metric call.
//
// @docspec:deterministic
func extractPrometheusLabels(call *ast.CallExpr) []string {
	sel, ok := call.Fun.(*ast.SelectorExpr)
	if !ok {
		return nil
	}

	// Vec metrics have a second argument with label names
	if !strings.HasSuffix(sel.Sel.Name, "Vec") || len(call.Args) < 2 {
		return nil
	}

	comp, ok := call.Args[1].(*ast.CompositeLit)
	if !ok {
		return nil
	}

	var labels []string
	for _, elt := range comp.Elts {
		if lit, ok := elt.(*ast.BasicLit); ok {
			labels = append(labels, strings.Trim(lit.Value, `"`))
		}
	}
	return labels
}

// extractStructField extracts a string literal value from a struct literal field.
//
// @docspec:deterministic
func extractStructField(comp *ast.CompositeLit, fieldName string) string {
	for _, elt := range comp.Elts {
		kv, ok := elt.(*ast.KeyValueExpr)
		if !ok {
			continue
		}
		ident, ok := kv.Key.(*ast.Ident)
		if !ok || ident.Name != fieldName {
			continue
		}
		lit, ok := kv.Value.(*ast.BasicLit)
		if !ok {
			continue
		}
		return strings.Trim(lit.Value, `"`)
	}
	return ""
}

// joinMetricName assembles a Prometheus metric name from namespace, subsystem, and name.
//
// @docspec:deterministic
func joinMetricName(namespace, subsystem, name string) string {
	parts := []string{}
	if namespace != "" {
		parts = append(parts, namespace)
	}
	if subsystem != "" {
		parts = append(parts, subsystem)
	}
	parts = append(parts, name)
	return strings.Join(parts, "_")
}

// isHealthCheckFunc checks if a function looks like a health check handler.
//
// @docspec:deterministic
func isHealthCheckFunc(fn *ast.FuncDecl) bool {
	name := strings.ToLower(fn.Name.Name)
	return strings.Contains(name, "health") || strings.Contains(name, "liveness") ||
		strings.Contains(name, "readiness") || strings.Contains(name, "ping")
}

// hasObservabilityDocTags checks for docspec observability comment tags.
//
// @docspec:deterministic
func hasObservabilityDocTags(file *ast.File) bool {
	for _, decl := range file.Decls {
		fn, ok := decl.(*ast.FuncDecl)
		if !ok || fn.Doc == nil {
			continue
		}
		for _, c := range fn.Doc.List {
			text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))
			if strings.HasPrefix(text, "docspec:metric") || strings.HasPrefix(text, "docspec:healthcheck") {
				return true
			}
		}
	}
	return false
}

File: rust/processor/src/extractor/observability.rs
//! Detect observability patterns in Rust projects.
//!
//! @docspec:module {
//!   id: "docspec-rust-extractor-observability",
//!   name: "Observability Extractor",
//!   description: "Detects tracing instrumentation, metric registrations (prometheus, metrics crate), and health check endpoints to populate the observability section of the DocSpec model.",
//!   since: "3.0.0"
//! }
//!
//! Looks for:
//! - `tracing` crate usage (`#[instrument]`, `tracing::info!`, span macros)
//! - `metrics` crate (counters, gauges, histograms)
//! - `prometheus` crate metric definitions
//! - `opentelemetry` usage
//! - Health check endpoints (common `/health`, `/healthz`, `/ready` patterns)

use serde_json::Value;
use quote::ToTokens;
use crate::scanner::FileInfo;
use super::{DocSpecExtractor, ProcessorContext};

/// Attribute names that indicate instrumentation / metrics.
const OBSERVABILITY_ATTRS: &[&str] = &[
    "instrument",      // tracing
    "tracing::instrument",
];

/// Metric registration patterns in source code.
const METRIC_PATTERNS: &[(&str, &str)] = &[
    ("counter!", "counter"),
    ("gauge!", "gauge"),
    ("histogram!", "histogram"),
    ("register_counter!", "counter"),
    ("register_gauge!", "gauge"),
    ("register_histogram!", "histogram"),
    ("Counter::new", "counter"),
    ("Gauge::new", "gauge"),
    ("Histogram::new", "histogram"),
    ("IntCounter::new", "counter"),
    ("IntGauge::new", "gauge"),
    ("HistogramVec::new", "histogram"),
];

/// Route paths that indicate health check endpoints.
const HEALTH_PATHS: &[&str] = &[
    "/health",
    "/healthz",
    "/ready",
    "/readiness",
    "/liveness",
    "/status",
    "/actuator/health",
];

/// @docspec:boundary "Observability detection across tracing, metrics, prometheus, and health check patterns"
pub struct ObservabilityExtractor;

impl DocSpecExtractor for ObservabilityExtractor {
    /// @docspec:intentional "Check if any observability crate (tracing, metrics, prometheus, opentelemetry) is in Cargo.toml"
    fn is_available(&self, ctx: &ProcessorContext) -> bool {
        ctx.has_dependency("tracing")
            || ctx.has_dependency("metrics")
            || ctx.has_dependency("prometheus")
            || ctx.has_dependency("opentelemetry")
    }

    /// @docspec:deterministic
    fn extractor_name(&self) -> &'static str {
        "observability"
    }

    /// @docspec:intentional "Extract tracing instrumentation, metric registrations, and health check endpoints from source"
    fn extract(&self, files: &[FileInfo], ctx: &mut ProcessorContext) {
        let mut metrics: Vec<Value> = Vec::new();
        let mut health_checks: Vec<Value> = Vec::new();
        let mut traced_functions: Vec<String> = Vec::new();

        for file_info in files {
            let syntax = match syn::parse_file(&file_info.source) {
                Ok(s) => s,
                Err(_) => continue,
            };

            for item in &syntax.items {
                match item {
                    syn::Item::Fn(func) => {
                        // Check for #[instrument]
                        if has_instrument_attr(&func.attrs) {
                            traced_functions.push(format!(
                                "{}::{}",
                                file_info.module,
                                func.sig.ident
                            ));
                        }

                        // Check for health check route paths
                        if let Some(path) = extract_health_path(&func.attrs) {
                            health_checks.push(serde_json::json!({
                                "path": path,
                                "checks": [func.sig.ident.to_string()],
                            }));
                        }

                        // Scan body for metric registrations
                        let func_block = &func.block;
                        let body = quote::quote!(#func_block).to_string();
                        extract_metrics(&body, &file_info.module, &func.sig.ident.to_string(), &mut metrics);
                    }
                    syn::Item::Impl(imp) => {
                        let owner = if let syn::Type::Path(tp) = imp.self_ty.as_ref() {
                            tp.path.segments.last()
                                .map(|s| s.ident.to_string())
                                .unwrap_or_default()
                        } else {
                            String::new()
                        };

                        for impl_item in &imp.items {
                            if let syn::ImplItem::Fn(method) = impl_item {
                                if has_instrument_attr(&method.attrs) {
                                    traced_functions.push(format!(
                                        "{}::{}#{}",
                                        file_info.module, owner, method.sig.ident
                                    ));
                                }

                                if let Some(path) = extract_health_path(&method.attrs) {
                                    health_checks.push(serde_json::json!({
                                        "path": path,
                                        "checks": [format!("{}::{}", owner, method.sig.ident)],
                                    }));
                                }

                                let method_block = &method.block;
                                let body = quote::quote!(#method_block).to_string();
                                let qualified = format!("{}::{}#{}", file_info.module, owner, method.sig.ident);
                                extract_metrics(&body, &qualified, &method.sig.ident.to_string(), &mut metrics);
                            }
                        }
                    }
                    // Detect static metric definitions (e.g., `static REQUESTS: Lazy<Counter> = ...`)
                    syn::Item::Static(s) => {
                        let static_ty = &s.ty;
                        let ty = quote::quote!(#static_ty).to_string();
                        let name = s.ident.to_string();
                        if let Some(metric_type) = detect_metric_type(&ty) {
                            metrics.push(serde_json::json!({
                                "name": name.to_lowercase().replace('_', "."),
                                "type": metric_type,
                                "emittedBy": [format!("{}::{}", file_info.module, name)],
                            }));
                        }
                    }
                    _ => {}
                }
            }
        }

        if metrics.is_empty() && health_checks.is_empty() && traced_functions.is_empty() {
            return;
        }

        let mut obs = serde_json::json!({});
        let map = obs.as_object_mut().unwrap();

        if !metrics.is_empty() {
            map.insert("metrics".to_string(), Value::Array(metrics));
        }
        if !health_checks.is_empty() {
            map.insert("healthChecks".to_string(), Value::Array(health_checks));
        }
        if !traced_functions.is_empty() {
            map.insert("tracedFunctions".to_string(),
                Value::Array(traced_functions.into_iter().map(Value::String).collect()));
        }

        ctx.observability = Some(obs);
    }
}

/// Check if an attribute list contains `#[instrument]` or `#[tracing::instrument]`.
///
/// @docspec:deterministic
fn has_instrument_attr(attrs: &[syn::Attribute]) -> bool {
    for attr in attrs {
        let path_str = attr.path().to_token_stream().to_string();
        for obs_attr in OBSERVABILITY_ATTRS {
            if path_str.contains(obs_attr) {
                return true;
            }
        }
    }
    false
}

/// Check if any route attribute points to a health check path.
///
/// @docspec:deterministic
fn extract_health_path(attrs: &[syn::Attribute]) -> Option<String> {
    let route_macros = ["get", "post", "route"];
    for attr in attrs {
        if let Some(ident) = attr.path().get_ident() {
            let name = ident.to_string();
            if route_macros.contains(&name.as_str()) {
                let tokens = attr.meta.to_token_stream().to_string();
                for health_path in HEALTH_PATHS {
                    if tokens.contains(health_path) {
                        return Some(health_path.to_string());
                    }
                }
            }
        }
    }
    None
}

/// Scan function body text for metric macro invocations.
///
/// @docspec:intentional "Scan function body for metric macro patterns (counter!, gauge!, histogram!) and extract metric names"
fn extract_metrics(body: &str, module: &str, fn_name: &str, metrics: &mut Vec<Value>) {
    for (pattern, metric_type) in METRIC_PATTERNS {
        if body.contains(pattern) {
            // Try to extract the metric name from the first string argument
            if let Some(pos) = body.find(pattern) {
                let after = &body[pos + pattern.len()..];
                if let Some(name) = extract_first_string(after) {
                    let already_exists = metrics.iter().any(|m| {
                        m.get("name").and_then(|n| n.as_str()) == Some(&name)
                    });
                    if !already_exists {
                        metrics.push(serde_json::json!({
                            "name": name,
                            "type": metric_type,
                            "emittedBy": [format!("{}::{}", module, fn_name)],
                        }));
                    }
                }
            }
        }
    }
}

/// Detect metric type from a static variable's type annotation.
///
/// @docspec:deterministic
fn detect_metric_type(ty: &str) -> Option<&'static str> {
    if ty.contains("Counter") || ty.contains("IntCounter") {
        Some("counter")
    } else if ty.contains("Gauge") || ty.contains("IntGauge") {
        Some("gauge")
    } else if ty.contains("Histogram") {
        Some("histogram")
    } else {
        None
    }
}

/// Extract the first quoted string from text.
///
/// @docspec:deterministic
fn extract_first_string(s: &str) -> Option<String> {
    let start = s.find('"')?;
    let rest = &s[start + 1..];
    let end = rest.find('"')?;
    let value = &rest[..end];
    if value.is_empty() {
        None
    } else {
        Some(value.to_string())
    }
}
